pages/donate.js

```diff
@@ -3,11 +3,137 @@ import Head from 'next/head';
 import Primary from '../components/Primary';
 import Sidebar from '../components/Sidebar';
 import DonationForm from '../components/forms/DonationForm';
+import ReviewForm from '../components/forms/ReviewForm';
 
 const pageTitle = 'Support TJI';
 
 class Page extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      formStep: 1,
+      firstName: '',
+      lastName: '',
+      email: '',
+      includeTax: false,
+      total: 0,
+      firstNameValid: false,
+      lastNameValid: false,
+      emailValid: false,
+      amountValid: false,
+      formValid: false,
+      amount: '0',
+      error: null,
+    };
+
+    this.handleInputChange = this.handleInputChange.bind(this);
+    this.submitForReview = this.submitForReview.bind(this);
+    this.returnToForm = this.returnToForm.bind(this);
+  }
+
+  // Handler for form inputs
+  handleInputChange = event => {
+    const { target } = event;
+    const { name } = target;
+    const value = target.type === 'checkbox' ? target.checked : target.value;
+
+    // Update our state and call field validation
+    this.setState(
+      {
+        [name]: value,
+      },
+      () => {
+        this.validateField(name, value);
+      }
+    );
+  };
+
+  // Check that our current field is field and update state accordingly
+  validateField(fieldName, value) {
+    let { firstNameValid, lastNameValid, emailValid, amountValid, includeTax, amount } = this.state;
+
+    // Compute the total donation
+    amount = parseFloat(amount);
+    amount.toFixed(2);
+    const total = includeTax === true ? amount + amount * 0.022 + 0.03 : amount;
+    total.toFixed(2);
+
+    switch (fieldName) {
+      case 'firstName':
+        firstNameValid = value.length > 0;
+        break;
+      case 'lastName':
+        lastNameValid = value.length > 0;
+        break;
+      case 'email':
+        emailValid = value.match(/^([\w.%+-]+)@([\w-]+\.)+([\w]{2,})$/i);
+        break;
+      case 'amount':
+        amountValid = value > 0;
+        break;
+      case 'includeTax':
+        includeTax = value;
+        break;
+      default:
+        break;
+    }
+
+    // Update our state and check if form is complete and valid
+    this.setState(
+      {
+        firstNameValid,
+        lastNameValid,
+        emailValid,
+        amountValid,
+        includeTax,
+        total,
+      },
+      () => {
+        this.validateForm();
+      }
+    );
+  }
+
+  // Form validation function
+  validateForm() {
+    const { state } = this;
+    const { firstNameValid, lastNameValid, emailValid, amountValid } = state;
+    this.setState({
+      formValid: firstNameValid && lastNameValid && emailValid && amountValid,
+    });
+  }
+
+  submitForReview(event) {
+    event.preventDefault();
+
+    const { formValid } = this.state;
+
+    const formStep = formValid ? 2 : 1;
+    const error = formValid ? null : 'Please complete the necessary fields before continuing.';
```

Comment:
Could our validation make it more clear what needs to be addressed? Maybe we could highlight and add a message to the fields that are invalid? Or we could make this message more descriptive.

Reply:
@nholden , I made some significant changes to the validation. I'm highlighting text/email input fields that are invalid, and added a message to the amount field in the event none is selected. Have a look and let me know what you think.